src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -192,3 +172,62 @@ where
         }
     }
 }
+
+/// Handle the zwp_virtual_keyboard_v1::keymap request.
+fn update_keymap<D>(data: &VirtualKeyboardUserData<D>, format: u32, fd: OwnedFd, size: usize)
+where
+    D: SeatHandler + 'static,
+{
+    // Only libxkbcommon compatible keymaps are supported.
+    if format != KeymapFormat::XkbV1 as u32 {
+        debug!("Unsupported keymap format: {format:?}");
+        return;
+    }
+
+    // Ignore potentially malicious requests.
+    if size > MAX_KEYMAP_SIZE {
+        debug!("Excessive keymap size: {size:?}");
+        return;
+    }
+
+    // Read entire keymap.
+    let mut keymap_buffer = vec![0; size];
+    let mut file = unsafe { File::from_raw_fd(fd.as_raw_fd()) };
+    if let Err(err) = file.read_exact(&mut keymap_buffer) {
+        debug!("Could not read keymap: {err}");
+        return;
+    }
```

Comment:
I know I'm coming after the merge, but I suppose we could have avoided that unsafe by just using `nix::unistd::read()` instead?

More generally, can creating this `File` do stupid stuff if a malicious client sends a stupid FD?

And writing that I realize that if the client sends an fd to something that is not a file/memory mapped region, but instead a pipe for example, it could block the compositor, no?

Reply:
> More generally, can creating this File do stupid stuff if a malicious client sends a stupid FD?

Yeah that's fair. The docs specifically state `The fd passed in must be a valid and open file descriptor`, so passing an invalid descriptor is just UB?

> And writing that I realize that if the client sends an fd to something that is not a file/memory mapped region, but instead a pipe for example, it could block the compositor, no?

I always forget under which circumstances the `WouldBlock` error is emitted, but if it isn't for pipes then that could be an issue, yeah.

Reply:
> so passing an invalid descriptor is just UB?

We cannot receive an invalid FD from the wayland socket, imo the main worry is what if the passed FD is to something else than a file or memory buffer.

About the use of `read()`, mostly my question was whether using `File::from_raw_fd()` implies more assumptions about the FD than just using `read()`, which would not be unsafe and just return an error if the FD cannot be read.

Reply:
> About the use of read(), mostly my question was whether using File::from_raw_fd() implies more assumptions about the FD than just using read(), which would not be unsafe and just return an error if the FD cannot be read.

Unfortunately I'm not aware of Rust internals, but even with `read()` you could get blocking.